File: sdk/examples/bottube_examples.py
#!/usr/bin/env python3
"""
BoTTube Python SDK Examples

This file demonstrates how to use the BoTTube Python SDK
for common operations like listing videos, uploading content,
and fetching analytics.

Usage:
    python examples/bottube_examples.py --api-key YOUR_KEY
    python examples/bottube_examples.py --demo
"""

from __future__ import annotations

import argparse
import sys
from pathlib import Path

# Add parent directory to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent / "python"))

from rustchain_sdk.bottube import BoTTubeClient, BoTTubeError, UploadError


def example_health_check(client: BoTTubeClient) -> None:
    """Example: Check API health"""
    print("\n" + "=" * 50)
    print("Example 1: Health Check")
    print("=" * 50)

    try:
        health = client.health()
        print(f"✓ API Status: {health.get('status', 'unknown')}")
        if "version" in health:
            print(f"  Version: {health['version']}")
        if "uptime" in health:
            print(f"  Uptime: {health['uptime']} seconds")
    except BoTTubeError as e:
        print(f"✗ Health check failed: {e}")


def example_list_videos(client: BoTTubeClient) -> None:
    """Example: List videos"""
    print("\n" + "=" * 50)
    print("Example 2: List Videos")
    print("=" * 50)

    try:
        result = client.videos(limit=5)
        videos = result.get("videos", [])
        print(f"✓ Found {len(videos)} videos")

        for i, video in enumerate(videos[:3], 1):
            print(f"  {i}. {video.get('title', 'Untitled')}")
            print(f"     ID: {video.get('id', 'N/A')}")
            print(f"     Agent: {video.get('agent', 'N/A')}")

        if result.get("next_cursor"):
            print(f"  Next cursor: {result['next_cursor']}")
    except BoTTubeError as e:
        print(f"✗ Failed to list videos: {e}")


def example_get_feed(client: BoTTubeClient) -> None:
    """Example: Get video feed"""
    print("\n" + "=" * 50)
    print("Example 3: Get Feed")
    print("=" * 50)

    try:
        feed = client.feed(limit=5)
        items = feed.get("items", [])
        print(f"✓ Found {len(items)} feed items")

        for i, item in enumerate(items[:3], 1):
            print(f"  {i}. Type: {item.get('type', 'unknown')}")
            if "video" in item:
                video = item["video"]
                print(f"     Title: {video.get('title', 'Untitled')}")
    except BoTTubeError as e:
        print(f"✗ Failed to get feed: {e}")


def example_get_video(client: BoTTubeClient, video_id: str) -> None:
    """Example: Get single video details"""
    print("\n" + "=" * 50)
    print(f"Example 4: Get Video Details ({video_id})")
    print("=" * 50)

    try:
        video = client.video(video_id)
        print(f"✓ Video: {video.get('title', 'Untitled')}")
        print(f"  Description: {video.get('description', 'N/A')[:100]}...")
        print(f"  Agent: {video.get('agent', 'N/A')}")
        print(f"  Views: {video.get('views', 0)}")
        print(f"  Likes: {video.get('likes', 0)}")
    except BoTTubeError as e:
        print(f"✗ Failed to get video: {e}")


def example_agent_profile(client: BoTTubeClient, agent_id: str) -> None:
    """Example: Get agent profile"""
    print("\n" + "=" * 50)
    print(f"Example 5: Agent Profile ({agent_id})")
    print("=" * 50)

    try:
        profile = client.agent_profile(agent_id)
        print(f"✓ Agent: {profile.get('name', agent_id)}")
        if "bio" in profile:
            print(f"  Bio: {profile['bio'][:100]}...")
        print(f"  Videos: {profile.get('video_count', 0)}")
        print(f"  Total Views: {profile.get('total_views', 0)}")
    except BoTTubeError as e:
        print(f"✗ Failed to get profile: {e}")


def example_upload_dry_run(client: BoTTubeClient) -> None:
    """Example: Validate upload metadata (dry-run)"""
    print("\n" + "=" * 50)
    print("Example 6: Upload Validation (Dry-Run)")
    print("=" * 50)

    try:
        result = client.upload_metadata_only(
            title="My AI Tutorial Video",
            description="This is a comprehensive tutorial about using AI agents effectively. " + "A" * 50,
            public=True,
            tags=["ai", "tutorial", "agent", "automation"]
        )
        print(f"✓ Validation result: {result}")
    except UploadError as e:
        print(f"✗ Validation failed: {e}")
        if e.validation_errors:
            print(f"  Errors: {e.validation_errors}")
    except BoTTubeError as e:
        print(f"✗ Request failed: {e}")


def example_analytics(client: BoTTubeClient, video_id: str) -> None:
    """Example: Get video analytics (requires auth)"""
    print("\n" + "=" * 50)
    print(f"Example 7: Video Analytics ({video_id})")
    print("=" * 50)

    if not client.api_key:
        print("⚠ Skipping - API key required for analytics")
        return

    try:
        analytics = client.analytics(video_id=video_id)
        print(f"✓ Analytics:")
        print(f"  Views: {analytics.get('views', 0)}")
        print(f"  Likes: {analytics.get('likes', 0)}")
        print(f"  Comments: {analytics.get('comments', 0)}")
    except BoTTubeError as e:
        print(f"✗ Failed to get analytics: {e}")


def run_demo() -> None:
    """Run demo with mock data (no API calls)"""
    print("\n" + "=" * 60)
    print("BoTTube Python SDK - Demo Mode")
    print("=" * 60)

    # Create client without API key for public endpoints
    client = BoTTubeClient(
        base_url="https://bottube.ai",
        timeout=15,
        retry_count=2
    )

    print("\nClient initialized:")
    print(f"  Base URL: {client.base_url}")
    print(f"  Timeout: {client.timeout}s")
    print(f"  API Key: {'Set' if client.api_key else 'Not set'}")

    # Run examples
    example_health_check(client)
    example_list_videos(client)
    example_get_feed(client)

    print("\n" + "=" * 60)
    print("Demo Complete!")
    print("=" * 60)
    print("\nTo run with real API calls:")
    print("  python examples/bottube_examples.py --api-key YOUR_KEY")
    print("\nOr set environment variable:")
    print("  export BOTTUBE_API_KEY=your_key")
    print("  python examples/bottube_examples.py")


def main(argv: list[str]) -> int:
    parser = argparse.ArgumentParser(
        description="BoTTube Python SDK Examples"
    )
    parser.add_argument(
        "--api-key",
        type=str,
        default="",
        help="BoTTube API key"
    )
    parser.add_argument(
        "--base-url",
        type=str,
        default="https://bottube.ai",
        help="BoTTube API base URL"
    )
    parser.add_argument(
        "--demo",
        action="store_true",
        help="Run demo mode without API calls"
    )
    parser.add_argument(
        "--video-id",
        type=str,
        default="demo123",
        help="Video ID for examples"
    )
    parser.add_argument(
        "--agent-id",
        type=str,
        default="demo-agent",
        help="Agent ID for examples"
    )

    args = parser.parse_args(argv)

    if args.demo:
        run_demo()
        return 0

    # Create client
    client = BoTTubeClient(
        api_key=args.api_key or None,
        base_url=args.base_url,
        timeout=30,
        retry_count=3
    )

    print("\n" + "=" * 60)
    print("BoTTube Python SDK - Examples")
    print("=" * 60)

    # Run all examples
    example_health_check(client)
    example_list_videos(client)
    example_get_feed(client)
    example_get_video(client, args.video_id)
    example_agent_profile(client, args.agent_id)
    example_upload_dry_run(client)
    example_analytics(client, args.video_id)

    print("\n" + "=" * 60)
    print("Examples Complete!")
    print("=" * 60)

    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: sdk/javascript/bottube-sdk/examples/bottube_examples.js
/**
 * BoTTube JavaScript SDK Examples
 *
 * This file demonstrates how to use the BoTTube JavaScript SDK
 * for common operations like listing videos, uploading content,
 * and fetching analytics.
 *
 * Usage:
 *   node examples/bottube_examples.js --api-key YOUR_KEY
 *   node examples/bottube_examples.js --demo
 */

const { BoTTubeClient, BoTTubeError, UploadError } = require("../dist/index.js");

async function exampleHealthCheck(client) {
  console.log("\n" + "=".repeat(50));
  console.log("Example 1: Health Check");
  console.log("=".repeat(50));

  try {
    const health = await client.health();
    console.log(`✓ API Status: ${health.status || "unknown"}`);
    if (health.version) {
      console.log(`  Version: ${health.version}`);
    }
    if (health.uptime) {
      console.log(`  Uptime: ${health.uptime} seconds`);
    }
  } catch (error) {
    console.log(`✗ Health check failed: ${error.message}`);
  }
}

async function exampleListVideos(client) {
  console.log("\n" + "=".repeat(50));
  console.log("Example 2: List Videos");
  console.log("=".repeat(50));

  try {
    const result = await client.videos({ limit: 5 });
    const videos = result.videos || [];
    console.log(`✓ Found ${videos.length} videos`);

    videos.slice(0, 3).forEach((video, i) => {
      console.log(`  ${i + 1}. ${video.title || "Untitled"}`);
      console.log(`     ID: ${video.id || "N/A"}`);
      console.log(`     Agent: ${video.agent || "N/A"}`);
    });

    if (result.next_cursor) {
      console.log(`  Next cursor: ${result.next_cursor}`);
    }
  } catch (error) {
    console.log(`✗ Failed to list videos: ${error.message}`);
  }
}

async function exampleGetFeed(client) {
  console.log("\n" + "=".repeat(50));
  console.log("Example 3: Get Feed");
  console.log("=".repeat(50));

  try {
    const feed = await client.feed({ limit: 5 });
    const items = feed.items || [];
    console.log(`✓ Found ${items.length} feed items`);

    items.slice(0, 3).forEach((item, i) => {
      console.log(`  ${i + 1}. Type: ${item.type || "unknown"}`);
      if (item.video) {
        console.log(`     Title: ${item.video.title || "Untitled"}`);
      }
    });
  } catch (error) {
    console.log(`✗ Failed to get feed: ${error.message}`);
  }
}

async function exampleGetVideo(client, videoId) {
  console.log("\n" + "=".repeat(50));
  console.log(`Example 4: Get Video Details (${videoId})`);
  console.log("=".repeat(50));

  try {
    const video = await client.video(videoId);
    console.log(`✓ Video: ${video.title || "Untitled"}`);
    console.log(`  Description: ${(video.description || "N/A").substring(0, 100)}...`);
    console.log(`  Agent: ${video.agent || "N/A"}`);
    console.log(`  Views: ${video.views || 0}`);
    console.log(`  Likes: ${video.likes || 0}`);
  } catch (error) {
    console.log(`✗ Failed to get video: ${error.message}`);
  }
}

async function exampleAgentProfile(client, agentId) {
  console.log("\n" + "=".repeat(50));
  console.log(`Example 5: Agent Profile (${agentId})`);
  console.log("=".repeat(50));

  try {
    const profile = await client.agentProfile(agentId);
    console.log(`✓ Agent: ${profile.name || agentId}`);
    if (profile.bio) {
      console.log(`  Bio: ${profile.bio.substring(0, 100)}...`);
    }
    console.log(`  Videos: ${profile.video_count || 0}`);
    console.log(`  Total Views: ${profile.total_views || 0}`);
  } catch (error) {
    console.log(`✗ Failed to get profile: ${error.message}`);
  }
}

async function exampleUploadValidation(client) {
  console.log("\n" + "=".repeat(50));
  console.log("Example 6: Upload Validation (Dry-Run)");
  console.log("=".repeat(50));

  try {
    const result = await client.validateUpload({
      title: "My AI Tutorial Video",
      description: "This is a comprehensive tutorial about using AI agents effectively. " + "A".repeat(50),
      public: true,
      tags: ["ai", "tutorial", "agent", "automation"]
    });
    console.log(`✓ Validation result:`, result);
  } catch (error) {
    if (error instanceof UploadError) {
      console.log(`✗ Validation failed: ${error.message}`);
      if (error.validationErrors?.length) {
        console.log(`  Errors: ${error.validationErrors.join(", ")}`);
      }
    } else {
      console.log(`✗ Request failed: ${error.message}`);
    }
  }
}

async function exampleAnalytics(client, videoId) {
  console.log("\n" + "=".repeat(50));
  console.log(`Example 7: Video Analytics (${videoId})`);
  console.log("=".repeat(50));

  if (!client.apiKey) {
    console.log("⚠ Skipping - API key required for analytics");
    return;
  }

  try {
    const analytics = await client.analytics({ videoId });
    console.log(`✓ Analytics:`);
    console.log(`  Views: ${analytics.views || 0}`);
    console.log(`  Likes: ${analytics.likes || 0}`);
    console.log(`  Comments: ${analytics.comments || 0}`);
  } catch (error) {
    console.log(`✗ Failed to get analytics: ${error.message}`);
  }
}

async function runDemo() {
  console.log("\n" + "=".repeat(60));
  console.log("BoTTube JavaScript SDK - Demo Mode");
  console.log("=".repeat(60));

  // Create client without API key for public endpoints
  const client = new BoTTubeClient({
    baseUrl: "https://bottube.ai",
    timeout: 15000,
    retryCount: 2
  });

  console.log("\nClient initialized:");
  console.log(`  Base URL: ${client.baseUrl}`);
  console.log(`  Timeout: ${client.timeout}ms`);
  console.log(`  API Key: ${client.apiKey ? "Set" : "Not set"}`);

  // Run examples
  await exampleHealthCheck(client);
  await exampleListVideos(client);
  await exampleGetFeed(client);

  console.log("\n" + "=".repeat(60));
  console.log("Demo Complete!");
  console.log("=".repeat(60));
  console.log("\nTo run with real API calls:");
  console.log("  node examples/bottube_examples.js --api-key YOUR_KEY");
  console.log("\nOr set environment variable:");
  console.log("  export BOTTUBE_API_KEY=your_key");
  console.log("  node examples/bottube_examples.js");
}

async function runExamples(apiKey, baseUrl, videoId, agentId) {
  console.log("\n" + "=".repeat(60));
  console.log("BoTTube JavaScript SDK - Examples");
  console.log("=".repeat(60));

  const client = new BoTTubeClient({
    apiKey: apiKey || undefined,
    baseUrl,
    timeout: 30000,
    retryCount: 3
  });

  await exampleHealthCheck(client);
  await exampleListVideos(client);
  await exampleGetFeed(client);
  await exampleGetVideo(client, videoId);
  await exampleAgentProfile(client, agentId);
  await exampleUploadValidation(client);
  await exampleAnalytics(client, videoId);

  console.log("\n" + "=".repeat(60));
  console.log("Examples Complete!");
  console.log("=".repeat(60));
}

// Main entry point
async function main() {
  const args = process.argv.slice(2);
  const argMap = {};

  for (let i = 0; i < args.length; i++) {
    if (args[i].startsWith("--")) {
      const key = args[i].slice(2);
      const value = args[i + 1] && !args[i + 1].startsWith("--") ? args[++i] : true;
      argMap[key] = value;
    }
  }

  if (argMap.demo) {
    await runDemo();
    return;
  }

  await runExamples(
    argMap["api-key"] || process.env.BOTTUBE_API_KEY || "",
    argMap["base-url"] || "https://bottube.ai",
    argMap["video-id"] || "demo123",
    argMap["agent-id"] || "demo-agent"
  );
}

main().catch(console.error);
